rpc, server: properly handle DRPC stream context cancellations

cthumuluru-crdb opened the pull request:
The graceful way to cancel a stream in DRPC is to close the stream. When stream context is canceled, DRPC closes the underlying transport by default as we see `<manager closed>` errors as a result. https://github.com/storj/drpc/blob/a5d487af8ae33deb7913b0f7c06b2c7ec7cd4dcc/drpcmanager/manager.go#L363-L377

`SoftClose` option on the DRPC manager allows for reusing the underlying transport as long as it's not busy (active write). In our case, since the stream is idle and at any point in time only one stream can use a connection we can use that option to avoid too many connection reopens. storj/drpc@a5d487a/drpcmanager/manager.go#L348-L363

Fixes: https://github.com/cockroachdb/cockroach/issues/140670
Epic: CRDB-48929
Release note: none

cthumuluru-crdb committed:
rpc, server: properly handle DRPC stream context cancellations

The graceful way to cancel a stream in DRPC is to close the stream. When stream context is canceled, DRPC
closes the underlying transport by default as we see `<manager closed>` errors as a result.
storj/drpc@a5d487a/drpcmanager/manager.go#L363-L378

`SoftClose` option on the DRPC manager allows for reusing the underlying transport as long as it's not busy
(active write). In our case, since the stream is idle and at any point in time only one stream can use a
connection we can use that option to avoid too many connection reopens.
storj/drpc@a5d487a/drpcmanager/manager.go#L348-L363

Fixes: https://github.com/cockroachdb/cockroach/issues/140670
Epic: CRDB-48929
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145907)
<!-- Reviewable:end -->

tbg reviewed (commented):
I've read through https://github.com/cockroachdb/cockroach/issues/140670#issuecomment-2845497525

If the 10s timeout hits in `runOnce` (say the conn is unused), isn't the goal to close the underlying connection (since there's a unique stream on top of the connection)? Aren't we leaking connections with this change? Concretely, what actually closes the underlying conn in `pooledStream.{close,Close}`?

For the third (and remaining) caller to `streamCancel()` this PR makes sense:

https://github.com/cockroachdb/cockroach/blob/c0ae46fccd3fb6264b9e8518e85a482e5a7f2256/pkg/rpc/stream_pool.go#L168-L171

Today this seems to cancel the entire stream if a `Send()` call has a context cancellation. `SoftClose` should "fix" this (unless the sending is still inflight), and in the way that makes sense. (Though: if the conn is still busy, i.e. the Send call hasn't gone out yet, won't we log the same ugly error into the logs?)

What I don't understand is why we ever get context cancellations during `Send()` in sysbench. Do you have an answer to that?

cthumuluru-crdb committed:
rpc: configure DPRC conn pool with idle timeout

DPRC pool holds a set of DRPC connections to a peer node. Until now, pool was not configired with
any connection timeout under the assumption that cancellnig a stream would also free its underlying
connection. However, our benchmarks where QPS fluctuates over the course of a run relevealed that
streams are continously opened and closed. Because each stream in the pool corresponds to its own
TCP connection (unlike gRPC, which multiplexes streams over a single connection), this constant
opening and closing of a connection is undesireable. To address this, the PR adds a connection
expiration mechanism so that idle connections are closed independently of the streams.

Fixes: #140670
Epic: CRDB-48929
Release note: none

cthumuluru-crdb commented on pkg/rpc/drpc.go:
```diff
@@ -110,6 +110,7 @@ func dialDRPC(rpcCtx *Context) func(ctx context.Context, target string) (drpcpoo
 					Stream: drpcstream.Options{
 						MaximumBufferSize: 0, // unlimited
 					},
+					SoftCancel: true, // Don't close the transport when stream context is canceled
```
Done
<!-- Reviewable comment -OPdWwajEf3ewkQ3y73U:bb6un1j -->
<!-- Sent from Reviewable.io -->

cthumuluru-crdb commented on pkg/server/drpc_test.go:
```diff
@@ -79,3 +80,76 @@ func TestDRPCBatchServer(t *testing.T) {
 	})
 }
+
+func TestStreamContextCancel(t *testing.T) {
```
I need to think about how to write such a test given the code is in two different packages (rpc, server). I'll track that separately to unblock this change unless you object to it. :)
<!-- Reviewable comment -OPdX0HW2L1vAe-lDchL:b-aknsmr -->
<!-- Sent from Reviewable.io -->

cthumuluru-crdb commented on pkg/server/drpc_test.go:
```diff
@@ -79,3 +80,76 @@ func TestDRPCBatchServer(t *testing.T) {
+	args := base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+		ServerArgs: base.TestServerArgs{
+			Insecure: true, // Secure connection is not required for this test
```
Done
<!-- Reviewable comment -OPdWyvH7-brR4hSyXpM:bb6un1j -->
<!-- Sent from Reviewable.io -->

cthumuluru-crdb reviewed (commented):
> If the 10s timeout hits in `runOnce` (say the conn is unused), isn't the goal to close the underlying connection (since there's a unique stream on top of the connection)?

Yes. That is correct. However, in the benchmark runs we see the streams open/close continuously with change in the QPS during the run. Steams are lighter to create and dispose but connections are relatively expensive. Without this change, every time the stream is closed, the underlying connection is closed. 

![image.png](https://files.reviewableusercontent.io/images/reviewable/176475514/-8Tl-V8TF0eVszD-wJt_zcwG/image.png)

 Aren't we leaking connections with this change? Concretely, what actually closes the underlying conn in `pooledStream.{close,Close}`

Yes. Without idle timeout on the connection pool, we leak the connections. I added changes to configure the pool to expire idle connections. This timeout is different from the stream pool timeout and will help keep the connections longer than the streams.

> (Though: if the conn is still busy, i.e. the Send call hasn't gone out yet, won't we log the same ugly error into the logs?)

Yes. Since we configured a callback to log DRPC server errors, if the connection is busy when the context is cancelled, we will see those errors.

 > What I don't understand is why we ever get context cancellations during `Send()` in sysbench. Do you have an answer to that?

What we see in the benchmarks is stream idle timeouts and not context cancellations.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145907#-:-OPdV2afE0x82knOaCrs:bsgj4ng)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
👍🏽 LGTM

cthumuluru-crdb commented:
bors r+
